modules/programs/firefox.nix

```
@@ -233,37 +257,78 @@ in {
             };
 
             bookmarks = mkOption {
-              type = types.attrsOf (types.submodule ({ config, name, ... }: {
-                options = {
-                  name = mkOption {
-                    type = types.str;
-                    default = name;
-                    description = "Bookmark name.";
-                  };
-
-                  keyword = mkOption {
-                    type = types.nullOr types.str;
-                    default = null;
-                    description = "Bookmark search keyword.";
+              type = let
+                bookmarkSubmodule = types.submodule ({ config, name, ... }: {
+                  options = {
+                    name = mkOption {
+                      type = types.str;
+                      default = name;
+                      description = "Bookmark name.";
+                    };
+
+                    keyword = mkOption {
+                      type = types.nullOr types.str;
+                      default = null;
+                      description = "Bookmark search keyword.";
+                    };
+
+                    url = mkOption {
+                      type = types.str;
+                      description = "Bookmark url, use %s for search terms.";
+                    };
                   };
+                }) // {
+                  description = "bookmark submodule";
+                };
 
-                  url = mkOption {
-                    type = types.str;
-                    description = "Bookmark url, use %s for search terms.";
+                bookmarkType = types.addCheck bookmarkSubmodule (x: x ? "url");
+
+                directoryType = types.submodule ({ config, name, ... }: {
+                  options = {
+                    name = mkOption {
+                      type = types.str;
+                      default = name;
+                      description = "Directory name.";
+                    };
+
+                    bookmarks = mkOption {
+                      type = types.listOf bookmarkType;
+                      default = [ ];
+                      description = "Bookmarks within directory.";
+                    };
                   };
+                }) // {
+                  description = "directory submodule";
                 };
-              }));
-              default = { };
+              in with types;
+              either (attrsOf bookmarkType)
```

To simplify handling both attribute sets and lists for the `bookmarks` option, `lib.coercedTo` could be used here to convert attribute sets to lists:
```suggestion

                nodeType = types.either bookmarkType directoryType;
              in with types;
              coercedTo (attrsOf nodeType) attrValues (listOf nodeType);
```
Doing this means that `bookmarks` does not have to be converted when calling `allItemsToHTML` nor checked against both an empty list *and* set (i.e. `bookmarks != [ ]` could be used instead of `bookmarks != [ ] && bookmarks != { }`).